[Opt Overlap] Implement _OverlappedOptimizer

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #71620
* #71608
* #71607
* #71606
* __->__ #71605
* #71604
* #71603
* #71602

Implements the main `_OverlappedOptimizer` as designed in https://github.com/pytorch/pytorch/issues/67570:

- `register_overlapped` decorator that allows an `_OverlappedOptimizer` to register the types of optimizers it supports overlap for
-  `_OverlappedStandardOptimizer` to overlap a standard `torch.optim.Optimizer`.
- `register_ddp` function to register the communication hook with DDP
- `_as_overlapped_optim` to return an `OverlappedOptimizer` that supports overlapping optimizer for the given optimizer type.

Example of usage:

```
overlapped_optimizer = _as_overlapped_optim(optim_cls, *functional_optim_args, **functional_optim_kwargs)
overlapped_optimizer.register_ddp(ddp_model_with_optimizer_hook)
```

However, note that the user is not expected to do this, and this will be called by libraries such as DDP and FSDP to register overlapped optimizers.

Differential Revision: [D33692686](https://our.internmc.facebook.com/intern/diff/D33692686/)

rohan-varma committed:
[Opt Overlap] Implement _OverlappedOptimizer



Differential Revision: [D33692686](https://our.internmc.facebook.com/intern/diff/D33692686/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Opt Overlap] Implement _OverlappedOptimizer"


Implements the main `_OverlappedOptimizer` as designed in https://github.com/pytorch/pytorch/issues/67570:

- `register_overlapped` decorator that allows an `_OverlappedOptimizer` to register the types of optimizers it supports overlap for
-  `_OverlappedStandardOptimizer` to overlap a standard `torch.optim.Optimizer`.
- `register_ddp` function to register the communication hook with DDP
- `_as_overlapped_optim` to return an `OverlappedOptimizer` that supports overlapping optimizer for the given optimizer type.

Example of usage:

```
overlapped_optimizer = _as_overlapped_optim(optim_cls, *functional_optim_args, **functional_optim_kwargs)
overlapped_optimizer.register_ddp(ddp_model_with_optimizer_hook)
```

However, note that the user is not expected to do this, and this will be called by libraries such as DDP and FSDP to register overlapped optimizers.

Differential Revision: [D33692686](https://our.internmc.facebook.com/intern/diff/D33692686/)

[ghstack-poisoned]

cbalioglu reviewed (approved):
Neat work! Left a comment just for future reference. Looks good to me.

rohan-varma committed:
Update on "[Opt Overlap] Implement _OverlappedOptimizer"


Implements the main `_OverlappedOptimizer` as designed in https://github.com/pytorch/pytorch/issues/67570:

- `register_overlapped` decorator that allows an `_OverlappedOptimizer` to register the types of optimizers it supports overlap for
-  `_OverlappedStandardOptimizer` to overlap a standard `torch.optim.Optimizer`.
- `register_ddp` function to register the communication hook with DDP
- `_as_overlapped_optim` to return an `OverlappedOptimizer` that supports overlapping optimizer for the given optimizer type.

Example of usage:

```
overlapped_optimizer = _as_overlapped_optim(optim_cls, *functional_optim_args, **functional_optim_kwargs)
overlapped_optimizer.register_ddp(ddp_model_with_optimizer_hook)
```

However, note that the user is not expected to do this, and this will be called by libraries such as DDP and FSDP to register overlapped optimizers.

Differential Revision: [D33692686](https://our.internmc.facebook.com/intern/diff/D33692686/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/algorithms/_optimizer_overlap/optimizer_overlap.py:
```diff
@@ -0,0 +1,84 @@
+    """
+    Returns a new ``OverlappedOptimizer`` instance that supports ``optim_cls``.
+    """
+    for registered_optim_cls, overlapped_optim_cls in _registered_overlapped_optims.items():
```
Awesome suggestion! I didn't think about issubclass() potentially not being O(1).

rohan-varma committed:
Update on "[Opt Overlap] Implement _OverlappedOptimizer"


Implements the main `_OverlappedOptimizer` as designed in https://github.com/pytorch/pytorch/issues/67570:

- `register_overlapped` decorator that allows an `_OverlappedOptimizer` to register the types of optimizers it supports overlap for
-  `_OverlappedStandardOptimizer` to overlap a standard `torch.optim.Optimizer`.
- `register_ddp` function to register the communication hook with DDP
- `_as_overlapped_optim` to return an `OverlappedOptimizer` that supports overlapping optimizer for the given optimizer type.

Example of usage:

```
overlapped_optimizer = _as_overlapped_optim(optim_cls, *functional_optim_args, **functional_optim_kwargs)
overlapped_optimizer.register_ddp(ddp_model_with_optimizer_hook)
```

However, note that the user is not expected to do this, and this will be called by libraries such as DDP and FSDP to register overlapped optimizers.

Differential Revision: [D33692686](https://our.internmc.facebook.com/intern/diff/D33692686/)

[ghstack-poisoned]
